### ISSUE TITLE
Randoop doesn't catch expected NullPointerException when generating regression tests

### ISSUE THREAD

**Issue by rjust** [randoop/randoop#69]
Consider the following method (taken from Apache commons lang):
```
    /**                                                                   
     * ...       
     * @param source A <code>String</code>, part of which should be parsed.      
     * @param pos A <code>ParsePosition</code> object with index and error index information as described above.                                                       
     * @return A <code>Date</code> parsed from the string. In case of error, returns null. 
     * @throws NullPointerException if source or pos is null.                      
     */                                                                          
    public Date parse(String source, ParsePosition pos)
```

Randoop (when executed with `--forbid-null=false`) generates a regression test that triggers the expected NullPointerException. However, the generated test doesn't catch this expected exception, and consequently fails on every run.

My understanding is that `--forbid-null=false` disables the "No null pointer exceptions" contract, and I therefore expected the NullPointerException to be caught -- just as Randoop catches, for example, an IllegalArgumentException thrown during test generation.

Thanks,
René

**Comment by bjkeller** [comment]
I'm confused by your issue description. Are you asking to add behavior to the forbid-null flag? It does exactly what the manual describes.

**Comment by mernst** [comment]
Regardless of anything else, Randoop shouldn't be generating tests that fail.  If test code throws a null pointer exception, then Randoop probably shouldn't output that test; but if Randoop does output that test, then the exception should be caught so that the Randoop-generated test suite passes.

**Comment by rjust** [comment]
Randoop generates two types of test cases:
1) error-revealing tests (using [pre-defined contracts](https://rawgit.com/randoop/randoop/master/doc/index.html#error_revealing_tests))
2) regression tests

Regarding 1), one particular contract is that no `NullPointerException` is thrown if `null` inputs are forbidden, and I think it makes sense for Randoop to output tests that fail due to a violation of a contract. Note that in contrast to the other [pre-defined contracts](https://rawgit.com/randoop/randoop/master/doc/index.html#error_revealing_tests), a test that fails due to a `NullPointerException` might still be a false positive, though.

However, I don't think that Randoop should ever output a test that fails for a reason unrelated to the **enabled**, [pre-defined contracts](https://rawgit.com/randoop/randoop/master/doc/index.html#error_revealing_tests). Given that the command-line switch `--forbid-null=false` **disables** the "No null pointer exceptions" contract, Randooop should catch every `NullPointerException` in this case.

**Comment by rjust** [comment]
It seems that this issue is related to the issue #20.

**Comment by bjkeller** [comment]
Just for clarification, this is the documentation for --forbid-null: 

> **--forbid-null=boolean**. Never use null as input to methods or constructors. This option causes Randoop to abandon the method call rather than providing null as an input, when no non-null value of the appropriate type is available.

I agree that having tests that throw exceptions is not very useful, but this doesn't say anything about a contract, and the code does *exactly* what this says.

**Comment by rjust** [comment]
Hi Ben,

I was referring to the [Error-revealing Tests](https://rawgit.com/randoop/randoop/master/doc/index.html#error_revealing_tests) section in the user manual:

> **What kinds of errors does Randoop check for?**
> Currently, Randoop checks for a small set of contracts:

> *Equals to null*: o.equals(null) should return false
> *Reflexivity of equality*: o.equals(o) should return true
> *Symmetry of equality*: o1.equals(o2) implies o2.equals(o1)
> *Equals-hashcode*: If o1.equals(o2)==true, then o1.hashCode() == o2.hashCode()
> *No null pointer exceptions*: No NullPointerException is thrown if no null inputs are used in a test.

My interpretation of the last contract is: if `null` inputs are used in a test (behavior that is controlled by `--forbid-null` and `--null-ratio`), then this contract does not apply.

Thanks,
René

**Comment by bjkeller** [comment]
Thanks. I understand better now.  You should definitely read the "if" in the last statement as "as long as".  I don't believe there is any intended statement about what happens when a `null` is given as an input.

**Comment by bjkeller** [comment]
Fixed in v2.1.0:

1. forbid-null and null-ratio behavior not changed, but we understand, and, hopefully, can explain it better. Setting forbid-null=true simply indicates that Randoop can use a null if it needs to in order to generate a test. This will occur only if it is unable to find another value for the argument. Null-ratio determines the frequency with which will use a null instead of another value.
2. For NullPointerExceptions, use --npe-on-null-input=EXPECTED when you want to indicate that a NullPointerException should be expected behavior when a test method uses null as an input. To specify behavior when non-null inputs are given within a method, use --npe-on-non-null-input. These both take the values INVALID, ERROR, and EXPECTED.
